### PR TITLE
Add joint and motor acceleration reading

### PR DESCRIPTION
**Pull request by isorrentino**

### CODE REVIEW

File: src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h [hunk 1]
```diff
@@ -241,6 +241,29 @@ class YarpSensorBridge : public ISensorBridge,
...
+     * Get joint acceleration in rad/s^2
+     * @param[in] jointName name of the joint
+     * @param[out] jointAcceleration joint acceleration in radians per second squared
+     * @param[out] receiveTimeInSeconds time at which the measurement was received
+     * @return true/false in case of success/failure
+     */
+    bool getJointAcceleration(const std::string& jointName,
+                          double& jointAcceleration,
+                          OptionalDoubleRef receiveTimeInSeconds = {}) final;
+
```
Comment (GiulioRomualdi) [inline src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h:254]: Could you keep the right indentation?

File: src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h [hunk 2]
```diff
@@ -241,6 +241,29 @@ class YarpSensorBridge : public ISensorBridge,
...
+     * Get all joints' accelerations in rad/s^2
+     * @param[out] parameter all joints' accelerations in radians per second squared
+     * @param[out] receiveTimeInSeconds time at which the measurement was received
+     * @warning the size is decided at the configuration and remains fixed,
+     * and internal checks must be done at the implementation level by the Derived class.
+     * This means that the user must pass a resized argument "jointAccelerations" to this method
+     * @return true/false in case of success/failure
+     */
+    bool getJointAccelerations(Eigen::Ref<Eigen::VectorXd> jointAccelerations,
+                            OptionalDoubleRef receiveTimeInSeconds = {}) final;
```
Comment (GiulioRomualdi) [inline src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h:265]: As above

File: src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h [hunk 3]
```diff
@@ -496,6 +519,29 @@ class YarpSensorBridge : public ISensorBridge,
...
+     * @param[out] receiveTimeInSeconds time at which the measurement was received
+     * @return true/false in case of success/failure
+     */
+    virtual bool getMotorAcceleration(const std::string& jointName,
+                                  double& motorAcceleration,
+                                  OptionalDoubleRef receiveTimeInSeconds = {}) final;
+
+    /**
+     * Get all motors' accelerations in rad/s
+     * @param[out] parameter all motors' accelerations in radians per second squared
```
Comment (prashanthr05) [inline src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h:535]:
```suggestion
     * @param[out] motorAccelerations all motors' accelerations in radians per second squared
```

File: src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h [hunk 4]
```diff
@@ -241,6 +241,29 @@ class YarpSensorBridge : public ISensorBridge,
...
+     * @param[out] receiveTimeInSeconds time at which the measurement was received
+     * @return true/false in case of success/failure
+     */
+    bool getJointAcceleration(const std::string& jointName,
+                          double& jointAcceleration,
+                          OptionalDoubleRef receiveTimeInSeconds = {}) final;
+
+    /**
+     * Get all joints' accelerations in rad/s^2
+     * @param[out] parameter all joints' accelerations in radians per second squared
```
Comment (prashanthr05) [inline src/RobotInterface/YarpImplementation/include/BipedalLocomotion/RobotInterface/YarpSensorBridge.h:257]:
```suggestion
     * @param[out] jointAccelerations all joints' accelerations in radians per second squared
```

File: src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp [hunk 5]
```diff
@@ -363,6 +363,46 @@ bool YarpSensorBridge::getJointVelocities(Eigen::Ref<Eigen::VectorXd> jointVeloc
     return true;
 }
 
+bool YarpSensorBridge::getJointAcceleration(const std::string& jointName,
+                                        double& jointAcceleration,
+                                        OptionalDoubleRef receiveTimeInSeconds)
```
Comment (prashanthr05) [inline src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp:368]: please correct the indentation.

File: src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp [hunk 6]
```diff
@@ -363,6 +363,46 @@ bool YarpSensorBridge::getJointVelocities(Eigen::Ref<Eigen::VectorXd> jointVeloc
...
+
+    if (receiveTimeInSeconds)
+        receiveTimeInSeconds.value().get()
+            = m_pimpl->controlBoardRemapperMeasures.receivedTimeInSeconds;
+
+    return true;
+}
+
+bool YarpSensorBridge::getJointAccelerations(Eigen::Ref<Eigen::VectorXd> jointAccelerations,
+                                         OptionalDoubleRef receiveTimeInSeconds)
```
Comment (prashanthr05) [inline src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp:390]: please correct the indentation.

File: src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp [hunk 7]
```diff
@@ -854,3 +894,53 @@ bool YarpSensorBridge::getMotorVelocities(Eigen::Ref<Eigen::VectorXd> motorVeloc
 
     return true;
 }
+
+bool YarpSensorBridge::getMotorAcceleration(const std::string& jointName,
+                                        double& motorAcceleration,
+                                        OptionalDoubleRef receiveTimeInSeconds)
```
Comment (prashanthr05) [inline src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp:900]: please correct the indentation.

File: src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp [hunk 8]
```diff
@@ -854,3 +894,53 @@ bool YarpSensorBridge::getMotorVelocities(Eigen::Ref<Eigen::VectorXd> motorVeloc
...
+
+    if (receiveTimeInSeconds)
+        receiveTimeInSeconds.value().get()
+            = m_pimpl->controlBoardRemapperMeasures.receivedTimeInSeconds;
+
+    return true;
+}
+
+bool YarpSensorBridge::getMotorAccelerations(Eigen::Ref<Eigen::VectorXd> motorAccelerations,
+                                         OptionalDoubleRef receiveTimeInSeconds)
```
Comment (prashanthr05) [inline src/RobotInterface/YarpImplementation/src/YarpSensorBridge.cpp:929]: please correct the indentation.

File: src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h [hunk 9]
```diff
@@ -629,6 +658,35 @@ class ISensorBridge
...
+    virtual bool getMotorAcceleration(const std::string& jointName,
+                                  double& motorAcceleration,
+                                  OptionalDoubleRef receiveTimeInSeconds = {})
+    {
+        return false;
+    };
+
+    /**
+     * Get all motors' accelerations in rad/s^2
+     * @param[out] parameter all motors' accelerations in radians per second squared
```
Comment (prashanthr05) [inline src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:678]:
```suggestion
     * @param[out] motorAccelerations all motors' accelerations in radians per second squared
```

File: src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h [hunk 10]
```diff
@@ -250,6 +250,35 @@ class ISensorBridge
...
+    virtual bool getJointAcceleration(const std::string& jointName,
+                                  double& jointAcceleration,
+                                  OptionalDoubleRef receiveTimeInSeconds = {})
+    {
+        return false;
+    };
+
+    /**
+     * Get all joints' accelerations in rad/s^2
+     * @param[out] parameter all joints' accelerations in radians per second squared
```
Comment (prashanthr05) [inline src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:269]:
```suggestion
     * @param[out] jointAccelerations all joints' accelerations in radians per second squared
```

**Review by prashanthr05 — APPROVED** [review]
Thanks, @isorrentino ! super minor comments only related to the aesthetics.

**Comment by isorrentino** [comment]
Ready again for review.

**Comment by GiulioRomualdi** [comment]
The failure is not related to this PR 😄 

`spdlog` uses `fmt` to format the output print. 12 hours ago the following commit has been merged https://github.com/Homebrew/homebrew-core/commit/d032f2fdb59a337cd06ad58b6011d1ffe5c4688b and the latest version of `fmt` (8.1.1) is now installed with brew. As explained `fmt` [here](https://fmt.dev/latest/api.html#udt) requires specializing `formatter<T>` to handle a custom type. 

As you can notice, [the action is failing](https://github.com/ami-iit/bipedal-locomotion-framework/runs/4759384846?check_suite_focus=true#step:19:55) while compiling the following line
https://github.com/ami-iit/bipedal-locomotion-framework/blob/0a758c518edf324a2db09e53286712d7cf5c26e8/src/TextLogging/src/Logger.cpp#L35


I think that in the nearest future we will have a similar issue also for the conda installation (cc @traversaro)

A possible solution is to implement a what explained [here](https://fmt.dev/latest/api.html#udt) (but I don't know if this approach is back-compatible with the _old_ fmt version installed by ubuntu 20.04

image 

Another possibility it's to modify directly this line:

https://github.com/ami-iit/bipedal-locomotion-framework/blob/0a758c518edf324a2db09e53286712d7cf5c26e8/src/TextLogging/src/Logger.cpp#L35

**Comment by GiulioRomualdi** [comment]
@isorrentino could you try to handle [it](https://github.com/ami-iit/bipedal-locomotion-framework/pull/492#issuecomment-1008736884) in another PR? If you don't have time I can do it

**Comment by GiulioRomualdi** [comment]
> @isorrentino could you try to handle [it](https://github.com/ami-iit/bipedal-locomotion-framework/pull/492#issuecomment-1008736884) in another PR? If you don't have time I can do it

#495 should fix the problem

**Comment by GiulioRomualdi** [comment]
Waiting for CI and then I will merge it

**Comment by traversaro** [comment]
> The failure is not related to this PR 😄

Related issue: https://github.com/robotology/robotology-superbuild/issues/991 .

**Pull request merged**